### DISCUSSION TITLE
Overlapping projects

### DISCUSSION THREAD

**Discussion by normvcr** [community/community#23819] [New to GitHub]
<p>Am preparing a couple of projects, P1 and P2, to host on GitHub.  Both projects are part of a larger project, P, and have source code in common.  For this reason, I’d like to host the big project, P, on GitHub, as well as both P1 and P2.  On the P1 and P2 pages, I will explain this situation, and I would like the P1 clone url to pull just the P1 source files out of the big P project , and the same for the P2 project.  The big P project can have the usual clone url, if that is what someone wants to do – get all the source code for both projects.</p>
<p>So, my question is how to set up the clone urls on the P1 and P2 pages do this?</p>

**Comment by lee-dohm** [comment]
<p>Welcome to the community, we’re glad you stopped by! :slight_smile:</p>
<p>I’m not certain I’m understanding exactly what it is you’re trying to do, so if you’ll bear with me as I ask for clarification we’ll see what we can do to help. It sounds like what you want is to have three repositories on GitHub:</p>
<ul>
<li><a href="https://github.com/normvcr/p">https://github.com/normvcr/p</a></li>
<li><a href="https://github.com/normvcr/p1">https://github.com/normvcr/p1</a></li>
<li><a href="https://github.com/normvcr/p2">https://github.com/normvcr/p2</a></li>
</ul>
<p>And if someone clones from <a href="https://github.com/normvcr/p">https://github.com/normvcr/p</a>, they should get <code>normvcr/p</code> with <code>p1</code> and <code>p2</code> inside it?</p>
<p>If I’m understanding correctly, you may be interested in this GitHub Blog post about using submodules:</p>
<aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/1038121/181079423-df13dcc7-006a-4d62-83e6-bffca6c10e8d.png" class="site-icon" width="32" height="32">
      <a href="https://github.blog/2016-02-01-working-with-submodules/" target="_blank" title="03:52AM - 02 February 2016">The GitHub Blog – 2 Feb 16</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/460;"><img src="https://user-images.githubusercontent.com/1038121/181079424-87374cfc-c4ce-4034-8c37-d639ec9dc740.jpeg" class="thumbnail" width="690" height="460" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/9/9f54f9356c325f66ad6377aedb3ba9a587c71d05_2_690x460.jpeg, https://aws1.discourse-cdn.com/github/optimized/2X/9/9f54f9356c325f66ad6377aedb3ba9a587c71d05_2_1035x690.jpeg 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/9/9f54f9356c325f66ad6377aedb3ba9a587c71d05_2_1380x920.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/9/9f54f9356c325f66ad6377aedb3ba9a587c71d05_2_10x10.png"></div>

<h3><a href="https://github.blog/2016-02-01-working-with-submodules/" target="_blank">Working with submodules - The GitHub Blog</a></h3>

<p>Eventually, any interesting software project will come to depend on another project, library, or framework. Git provides submodules to help with this. Submodules allow you to include or embed one or more repositories as a</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>It describes a situation very similar to what you’re talking about and how you can make it all work.</p>
<p>I hope that helps!</p>

**Comment by normvcr** [comment]
<p>Thanks for the pointer to the interesting article on submodules.  I don’t think submodules will do it for me, as p1 and p2 share source code in common, and I do not want to duplicate source code – this is why I want to have just one repository,p, with all the source code, and the other projects can do, say, sparse checkout.  But, this requires some setup.  The p1 project would provide the patterns needed for the p1 sparse checkout, but how to integrate this file of patterns with the clone url that is the current interface for getting hold of p1?</p>

**Comment by lee-dohm** [comment, marked as answer]
<aside class="quote no-group" data-username="normvcr" data-post="3" data-topic="115789">
<div class="title">
<div class="quote-controls"></div>
 normvcr:</div>
<blockquote>
<p>how to integrate this file of patterns with the clone url that is the current interface for getting hold of p1?</p>
</blockquote>
</aside>
<p>There isn’t a way to customize the clone user experience that’s provided by the GitHub UI, it is set up to only clone the entire repository. If you want people to use some other mechanism to clone your repo, you would have to document it in the <code>README</code> (or some other document in the repository) and point people to it that way.</p>

**Comment by normvcr** [comment]
<p>OK, thank you for your feedback and time.</p>
